[SPARK-19602][SQL] Support column resolution of fully qualified column name ( 3 part name)

skambha opened the pull request:
## What changes were proposed in this pull request?
The design details is attached to the JIRA issue [here](https://drive.google.com/file/d/1zKm3aNZ3DpsqIuoMvRsf0kkDkXsAasxH/view)

High level overview of the changes are: 
- Enhance the qualifier to be more than one string
- Add support to store the qualifier. Enhance the lookupRelation to keep the qualifier appropriately.
- Enhance the table matching column resolution algorithm to account for qualifier being more than a string.
- Enhance the table matching algorithm in UnresolvedStar.expand
- Ensure that we continue to support select t1.i1 from db1.t1

## How was this patch tested?
- New tests are added.
- Several test scenarios were added in a separate  [test pr 17067](https://github.com/apache/spark/pull/17067).  The tests that were not supported earlier are marked with TODO markers and those are now supported with the code changes here. 
- Existing unit tests ( hive, catalyst and sql) were run successfully.

AmplabJenkins commented:
Can one of the admins verify this patch?

skambha commented:
cc @gatorsmile, @cloud-fan   I'd really appreciate your review and comments. Thanks much.

cloud-fan commented on sql/core/src/test/resources/sql-tests/inputs/columnresolution-views.sql:
```diff
@@ -13,10 +13,8 @@ DROP VIEW view1;
 CREATE OR REPLACE GLOBAL TEMPORARY VIEW view1 as SELECT 1 as i1;
 SELECT * FROM global_temp.view1;
--- TODO: Support this scenario
 SELECT global_temp.view1.* FROM global_temp.view1;
```
I'm not sure if this is a valid case. We should treat `SELECT ... FROM ...` as a 2-levels look up: Level 1, in `FROM` statement, we specify the table name with an optional qualifier(database name). Level 2, in `SELECT` statement, we specify the column name with an optional qualifier(table name).

I don't think it makes sense to specify the database name for column name, the table lookup should be done in the `FROM` statement.

cloud-fan replied:
cc @gatorsmile

gatorsmile replied:
Yes, but @skambha did a comparison between different DB. Basically, what she tried to do is to make Spark SQL consistent with the others. See the design doc:
https://issues.apache.org/jira/secure/attachment/12854681/Design_ColResolution_JIRA19602.pdf

skambha replied:
Thanks @gatorsmile and @cloud-fan.  

Hi @cloud-fan,  
The use-case is to support fully qualified column name in queries to remove ambiguity in the naming.  

For e.g. `select db2.t1.i1 from db1.t1, db2.t1 where db1.t1.i2 = db2.t1.i3`

Today Spark does not support this.   The workaround is to create alias for the table.
 
Yes, the lookup for the table is coming from the ‘from’ clause. At a high level, changes include keeping track of the qualifier(including the database name) when the lookup happens and enhancing the column resolution logic in the analyzer to resolve fully qualified column name from a logical plan node’s children.  
There are more details in the design doc attached to the[ SPARK-19602](https://issues.apache.org/jira/browse/SPARK-19602)

Thanks.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

cloud-fan commented:
I agree it's a valid use case, do you wanna bring it up to date? sorry for the delay!

skambha commented:
sure.  Let me look into it.  Thanks.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:
```diff
@@ -88,12 +88,12 @@ case class UnresolvedAttribute(nameParts: Seq[String]) extends Attribute with Un
   override def dataType: DataType = throw new UnresolvedException(this, "dataType")
   override def nullable: Boolean = throw new UnresolvedException(this, "nullable")
-  override def qualifier: Option[String] = throw new UnresolvedException(this, "qualifier")
+  override def qualifier: Option[Seq[String]] = throw new UnresolvedException(this, "qualifier")
```
I think we can just use `Seq[String]`

skambha replied:
I followed the style that exists today. If we change here, then there are other places where it needs to change as well.  For e.g, in AttributeReference we initialize it with None today and now it will be a Seq.empty.  I am not sure if we want to create these objects or just leave it using  None by keeping the Option[]  

If you have a strong preference, I can update it.   I do have the changes locally but have not pushed that version.   Thanks.

skambha commented:
I have rebased and pushed the changes.  I ran the unit tests ( sql, catalyst and hive). 
Earlier, I was having issues running the hive test suite locally but that is resolved with the fix from SPARK-23275.  

Please take a look.   Thanks.

gatorsmile commented:
ok to test

SparkQA commented:
**[Test build #86899 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86899/testReport)** for PR 17185 at commit [`8206dc3`](https://github.com/apache/spark/commit/8206dc3bc2595507ba71e7f50fddeed0c3b16479).

SparkQA commented:
**[Test build #86899 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86899/testReport)** for PR 17185 at commit [`8206dc3`](https://github.com/apache/spark/commit/8206dc3bc2595507ba71e7f50fddeed0c3b16479).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86899/
Test PASSed.

mafernandez-stratio commented:
Any plans of merging this PR?

gatorsmile commented:
@mafernandez-stratio We plan to review and merge after 2.3 release. The target release will be 2.4

mafernandez-stratio commented:
@gatorsmile Great, thanks for the feedback!

gatorsmile commented:
ok to test

SparkQA commented:
**[Test build #93810 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93810/testReport)** for PR 17185 at commit [`8206dc3`](https://github.com/apache/spark/commit/8206dc3bc2595507ba71e7f50fddeed0c3b16479).

skambha committed:
Squashed commits :
    Changes to resolve fully qualified column attribute references from a logical plan node children

SparkQA commented:
**[Test build #93810 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93810/testReport)** for PR 17185 at commit [`8206dc3`](https://github.com/apache/spark/commit/8206dc3bc2595507ba71e7f50fddeed0c3b16479).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93810/
Test FAILed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:
```diff
@@ -262,17 +262,47 @@ abstract class Star extends LeafExpression with NamedExpression {
+   * Returns true if the nameParts match the qualifier of the attribute
+   *
+   * There are two checks: i) Check if the nameParts match the qualifier fully.
+   * E.g. SELECT db.t1.* FROM db1.t1   In this case, the nameParts is Seq("db1", "t1") and
```
what about `SELECT db1.t1.* FROM t1` while the current database is `db1`?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:
```diff
@@ -654,16 +654,19 @@ class SessionCatalog(
       if (db == globalTempViewManager.database) {
         globalTempViewManager.get(table).map { viewDef =>
-          SubqueryAlias(table, viewDef)
+          SubqueryAlias(alias, viewDef)
```
I think we can make `SubqueryAlias` take a `qualifier: Seq[String]` instead of `alias: String`.

skambha replied:
Here, SubqueryAlias takes in a AliasIdentifier and not a string. 

Also want to mention that: we considered somewhat similar to your suggestion - to  simply add a qualifier as a Seq[String] as a member of SubqueryAlias but that change would require touching more files.  

I had implemented that first (long time back now) and then after some offline discussions we favored this as this change is more localized and encapsulated.  @gatorsmile @cloud-fan

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:
```diff
@@ -121,14 +129,14 @@ abstract class Attribute extends LeafExpression with NamedExpression with NullIn
 case class Alias(child: Expression, name: String)(
     val exprId: ExprId = NamedExpression.newExprId,
-    val qualifier: Option[String] = None,
+    val qualifier: Option[Seq[String]] = None,
```
Again, I feel using `Seq[String]` directly can simplify a lot of code.

skambha replied:
I'll look into this.

skambha replied:
I took care of this change in the recent push.  Thanks.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:
```diff
@@ -120,22 +120,54 @@ abstract class LogicalPlan
   /**
    * Resolve the given `name` string against the given attribute, returning either 0 or 1 match.
```
can you briefly explain the new resolution logic? I feel it's a little convoluted now as we are more likely to be ambiguous.

skambha replied:
The resolution logic needs to account to compare the full qualifier.   If you are trying to resolve db1.t1.i1, the match needs to happen against the entire qualifier of the attribute.   

For e.g: 
```SQL
 select db1.t1.i1 from t1   // current database is db1  and lets say t1 has  columns i1, i2. 
```
When resolving db1.t1.i1,  against the following set of Input attributes.   
we will have the following attribute references:
- AttributeReference ( qualifier Seq(db1, t1)  and name is i1)
- AttributeReference ( qualifier Seq(db1, t1) and name is i2) 

So the new resolution logic will match the entire qualifier sequence and then match the column ( ie the attribute name)

skambha replied:
At a high level the changes include: 
* The column resolution needs to account for the qualifier that is now a sequence of string. 
* Enhance the lookupRelation to add the qualifier appropriately. Right now, the qualifier used by the LogicalPlan nodes representing the relation leafnodes use only the table name as the qualifier when no alias is provided. This needs to be updated to add the database name and the table name for permanent views and tables and global temporary views. 

 Full details of the design and the resolution logic is in the [doc here](https://drive.google.com/file/d/1zKm3aNZ3DpsqIuoMvRsf0kkDkXsAasxH/view) in Section 3.3 and 3.4. 

So, Section 2 in the design doc has a lot of scenarios that were tested against different db’s and Spark and the expected behavior with the design here.   

If there is a specific scenario you have in mind that would be valid but get flagged as ambiguous, can you please let me know.   I’d be glad to look into it.

skambha commented:
@gatorsmile , @cloud-fan,  just a quick comment,  I have been working on this and will respond soon.

skambha committed:
Rebase, bypass the map logic and add todos for self and add a testcase

skambha commented:
I rebased and found out that the resolution code in Logical plan has changed and it uses map lookup to do the matching. I have some ideas on how to incorporate the 3 part name with the map lookup logic.  

For now,  I have synced up and bypassed the new map logic and pushed the code so it is up to the latest so I can get the test cycle from GitHub and added todo’s for myself to incorporate the optimized lookup logic. 

@cloud-fan,  In the meantime, if you have any other comments please let me know.  Thanks.

SparkQA commented:
**[Test build #93904 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93904/testReport)** for PR 17185 at commit [`05d6c0f`](https://github.com/apache/spark/commit/05d6c0fc6139f8672aea3173da3d98c5cc4e1a29).

SparkQA commented:
**[Test build #93904 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93904/testReport)** for PR 17185 at commit [`05d6c0f`](https://github.com/apache/spark/commit/05d6c0fc6139f8672aea3173da3d98c5cc4e1a29).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93904/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:
```diff
@@ -262,17 +262,47 @@ abstract class Star extends LeafExpression with NamedExpression {
+    val qualifierList = attribute.qualifier.getOrElse(Seq.empty)
+
+    // match the qualifiers and nameParts
+    val matched = nameParts.corresponds(qualifierList)(resolver) match {
```
it's weird to use pattern match like this, maybe better to write
```
nameParts.corresponds(qualifierList)(resolver) || {
  if (nameParts.length == 1 && qualifierList.nonEmpty) {
    resolver(nameParts.head, qualifierList.last)
  } else {
    false
  }
}
```

skambha replied:
done. Thanks.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:
```diff
@@ -262,17 +262,47 @@ abstract class Star extends LeafExpression with NamedExpression {
+    val matched = nameParts.corresponds(qualifierList)(resolver) match {
+      case true => true
+      case false if (nameParts.length == 1 && qualifierList.nonEmpty) =>
+        // check if it matches the table portion of the qualifier
```
IIRC users can do `select structCol.* from tbl`, does it mean the `nameParts.head` may not be table name?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:
```diff
@@ -71,19 +71,27 @@ trait NamedExpression extends Expression {
    * 1. Empty Seq: when an attribute doesn't have a qualifier,
    *    e.g. top level attributes aliased in the SELECT clause, or column from a LocalRelation.
-   * 2. Single element: either the table name or the alias name of the table.
+   * 2. Seq with a Single element: either the table name or the alias name of the table.
```
3. a seq of 2 elements: database name and table name.

skambha replied:
done.

cloud-fan commented:
overall LGTM, my major concern is how to do O(1) lookup for the 3 part name

skambha commented:
Thanks @cloud-fan for the review. 

I am working on implementing an idea to get optimized lookup with 3part name.

skambha committed:
Integrate with the map lookup logic and add optimized lookup for 3 part name and remove my todos

SparkQA commented:
**[Test build #94160 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94160/testReport)** for PR 17185 at commit [`4e9c1bf`](https://github.com/apache/spark/commit/4e9c1bfede87060cd89f227e1891d9a0046c8df9).

skambha committed:
Remove extra space

skambha commented:
retest this please

SparkQA commented:
**[Test build #94162 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94162/testReport)** for PR 17185 at commit [`8931354`](https://github.com/apache/spark/commit/8931354b01297833bda1dbcff21d42f928448b3a).

SparkQA commented:
**[Test build #94160 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94160/testReport)** for PR 17185 at commit [`4e9c1bf`](https://github.com/apache/spark/commit/4e9c1bfede87060cd89f227e1891d9a0046c8df9).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94160/
Test FAILed.

SparkQA commented:
**[Test build #94162 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94162/testReport)** for PR 17185 at commit [`8931354`](https://github.com/apache/spark/commit/8931354b01297833bda1dbcff21d42f928448b3a).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94162/
Test FAILed.

skambha commented:
Looks like there is another change that has gone in that this PR conflicts with so the build fails. :( 
I will rebase again and push.

skambha commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:
```diff
@@ -262,17 +262,47 @@ abstract class Star extends LeafExpression with NamedExpression {
+   * attribute is Seq("a")
+   */
+  private def matchedQualifier(
+      attribute: Attribute,
```
The database information gets carried over based on the lookupRelation logic.  Details are in Section 3.3.3 in the design doc [here](https://drive.google.com/file/d/1zKm3aNZ3DpsqIuoMvRsf0kkDkXsAasxH/view)

skambha committed:
Resolve conflict after rebase

SparkQA commented:
**[Test build #94172 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94172/testReport)** for PR 17185 at commit [`90cd6d3`](https://github.com/apache/spark/commit/90cd6d33f59fdae16a3a386ed14cefb3f28d35a8).

skambha commented:
I rebased and ran the catalyst unit test suite only as a sanity test.  fwiw, I had run the sql and hive and catalyst suites earlier prior to this last rebase.

skambha commented:
The last push has changes that adds the map lookup for the 3 part name.  It implements a solution to address the lookup.. although I think there are more ways we can go about it and could possibly improve on it in the future. 

@cloud-fan, please take a look.  Thanks.

SparkQA commented:
**[Test build #94172 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94172/testReport)** for PR 17185 at commit [`90cd6d3`](https://github.com/apache/spark/commit/90cd6d33f59fdae16a3a386ed14cefb3f28d35a8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94172/
Test FAILed.

dilipbiswal commented:
retest this please

SparkQA commented:
**[Test build #94187 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94187/testReport)** for PR 17185 at commit [`90cd6d3`](https://github.com/apache/spark/commit/90cd6d33f59fdae16a3a386ed14cefb3f28d35a8).

SparkQA commented:
**[Test build #94187 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94187/testReport)** for PR 17185 at commit [`90cd6d3`](https://github.com/apache/spark/commit/90cd6d33f59fdae16a3a386ed14cefb3f28d35a8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94187/
Test FAILed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:
```diff
@@ -316,8 +345,8 @@ case class UnresolvedRegex(regexPattern: String, table: Option[String], caseSens
       // If there is a table, pick out attributes that are part of this table that match expr
-      case Some(t) => input.output.filter(_.qualifier.exists(resolver(_, t)))
-        .filter(_.name.matches(pattern))
+      case Some(t) => input.output.filter(a => resolver(a.qualifier.last, t)).
```
do we need to consider if `qualifier` is Nil?

skambha replied:
Sure. Will add a defensive check.

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #94206 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94206/testReport)** for PR 17185 at commit [`90cd6d3`](https://github.com/apache/spark/commit/90cd6d33f59fdae16a3a386ed14cefb3f28d35a8).

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/package.scala:
```diff
@@ -169,25 +181,50 @@ package object expressions  {
+      //
+      // For example, consider an example where "db1" is the database name, "a" is the table name
+      // and "b" is the column name and "c" is the struct field name.
+      // If the name parts is db1.a.b.c, then Attribute will match
```
we may need to check the SQL standard about how to resolve the ambiguity. For `a.b.c`, it's possible that `a` is db name, `b` is table name, `c` is column name. If there is no column named `c`, we should fail. It's also possible that `a` is table name, `b` is column name, `c` is nested field name, and it exists. What's the expected behavior here?

skambha replied:
In this case, if a.b.c fails to resolve as db.table.column, then we check if there is a table  and column that matches a.b and then see if c is a nested field name and if it exists, it will resolve to the nested field. 

Tests with struct nested fields are [here](https://github.com/skambha/spark/blob/90cd6d33f59fdae16a3a386ed14cefb3f28d35a8/sql/core/src/test/resources/sql-tests/inputs/columnresolution.sql#L65-L73)

cloud-fan replied:
Is this behavior defined by SQL standard? IIRC the current behavior without your patch is
`col.innerField1.innerField2` can fail even we have a table named `col` and a column named `innerField1` which has a `innerField2` field.

skambha replied:
Actually it won't fail without this patch too,  the col.innerField1.innerField2 will resolve correctly to the struct field if there is a table named col and column named innerField1 which has a innerField2 field.   

Please see an example of such a scenario in this test output on master spark: 
https://github.com/apache/spark/blob/master/sql/core/src/test/resources/sql-tests/results/columnresolution.sql.out#L360
------
Let me look into the sql standard and get back on that.

cloud-fan replied:
What I'm talking about is ambiguity. `col.innerField1.innerField2` can fail if the `innerField2` doesn't exist. My question is: shall we try all the possible resolution paths and pick the valid one, or define a rule that we can decide the resolution path ahead.

skambha replied:
@cloud-fan , Thank you for your suggestion and question. 

Existing spark behavior follows precedence rules in the column resolution logic and in this patch we are following the same pattern/rule.   

I am looking into the SQL standard to see if there are any column resolution rules but I have not found any yet.  However when I researched existing databases, I observed different behaviors among them and it is listed in Section 2/Table A in the design doc [here](https://drive.google.com/file/d/1zKm3aNZ3DpsqIuoMvRsf0kkDkXsAasxH/view).

I agree, we can improve upon the checks in existing precedence to go all the way to ensure there is a nested field.  Although, the user can always qualify the field to resolve the ambiguity.  Shall we open another issue to discuss and improve upon the existing resolution logic.

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalogSuite.scala:
```diff
@@ -536,12 +536,13 @@ abstract class SessionCatalogSuite extends AnalysisTest {
       assert(metadata.viewText.isDefined)
       val view = View(desc = metadata, output = metadata.schema.toAttributes,
         child = CatalystSqlParser.parsePlan(metadata.viewText.get))
+      val alias = AliasIdentifier("view1", Some("db3"))
```
instead of creating `AliasIdentifier` manually, we can also define a `SubqueryAlias.apply` which takes db name and table name.

skambha replied:
Sure, let me do that if this style is preferred. Thanks.

cloud-fan reviewed (commented):
mostly LGTM

skambha committed:
address review comments

skambha commented:
I have addressed the  review comments in this commit [here](https://github.com/apache/spark/pull/17185/commits/065687f3b987e254f412796666d45f0cced6e42e)

@cloud-fan, please take a look.  Thanks.

SparkQA commented:
**[Test build #94234 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94234/testReport)** for PR 17185 at commit [`065687f`](https://github.com/apache/spark/commit/065687f3b987e254f412796666d45f0cced6e42e).

SparkQA commented:
**[Test build #94234 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94234/testReport)** for PR 17185 at commit [`065687f`](https://github.com/apache/spark/commit/065687f3b987e254f412796666d45f0cced6e42e).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94234/
Test FAILed.

dilipbiswal commented:
retest this please

SparkQA commented:
**[Test build #94244 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94244/testReport)** for PR 17185 at commit [`065687f`](https://github.com/apache/spark/commit/065687f3b987e254f412796666d45f0cced6e42e).

SparkQA commented:
**[Test build #94244 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94244/testReport)** for PR 17185 at commit [`065687f`](https://github.com/apache/spark/commit/065687f3b987e254f412796666d45f0cced6e42e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94244/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:
```diff
@@ -201,7 +204,7 @@ case class Alias(child: Expression, name: String)(
 
   override def sql: String = {
-    val qualifierPrefix = qualifier.map(_ + ".").getOrElse("")
+    val qualifierPrefix = if (qualifier.nonEmpty) qualifier.mkString(".") + "." else ""
```
nit: `val qualifierPrefix = qualifier.mkString("", ".", ".")`

skambha replied:
This won't work for the case when we have Seq.empty.  The suffix "." gets returned even for a empty sequence. 
For a non empty Seq, the above call will be fine. 

Shall we leave the 'if' as is  or is there an equivalent preferred style that would work?

cloud-fan replied:
ah my bad, I thought it would return empty string for empty seq. Let's leave it.

skambha replied:
ok, sounds good.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/package.scala:
```diff
@@ -169,25 +181,50 @@ package object expressions  {
-          val key = (qualifier.toLowerCase(Locale.ROOT), name.toLowerCase(Locale.ROOT))
-          val attributes = collectMatches(name, qualified.get(key)).filter { a =>
-            resolver(qualifier, a.qualifier.get)
+      matches = matches match {
```
nit:
```
if (matches._1.isEmpty) {
  matches = ...
}
```

skambha replied:
done.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:
```diff
@@ -794,19 +795,37 @@ case class LocalLimit(limitExpr: Expression, child: LogicalPlan) extends OrderPr
+
+  def apply(
+      identifier: String,
+      database: Option[String],
```
why it's `Option`? if no database name, we should call the `apply` method above

skambha replied:
good point! I'll take care of this in the next push.

cloud-fan commented:
LGTM except some code style comment, thanks for working on it!

skambha committed:
Address review comments - style related

SparkQA commented:
**[Test build #94335 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94335/testReport)** for PR 17185 at commit [`5f7e5d7`](https://github.com/apache/spark/commit/5f7e5d7bddca593d72818b07d71f678bd0a1982d).

skambha commented:
Thanks for the review.  I have addressed your comments and pushed the changes.  
@cloud-fan, Please take a look.

SparkQA commented:
**[Test build #94335 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94335/testReport)** for PR 17185 at commit [`5f7e5d7`](https://github.com/apache/spark/commit/5f7e5d7bddca593d72818b07d71f678bd0a1982d).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94335/
Test FAILed.

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #94353 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94353/testReport)** for PR 17185 at commit [`5f7e5d7`](https://github.com/apache/spark/commit/5f7e5d7bddca593d72818b07d71f678bd0a1982d).

SparkQA commented:
**[Test build #94353 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94353/testReport)** for PR 17185 at commit [`5f7e5d7`](https://github.com/apache/spark/commit/5f7e5d7bddca593d72818b07d71f678bd0a1982d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94353/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

gatorsmile commented:
@skambha @dilipbiswal How about the hint resolution after supporting multi part names?

gatorsmile commented:
Let us discuss it in the JIRA https://issues.apache.org/jira/browse/SPARK-25121

dilipbiswal commented:
Good idea @gatorsmile . Thanks !!
